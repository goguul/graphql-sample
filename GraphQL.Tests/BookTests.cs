using GraphQL.Books;
using GraphQL.Data;
using HotChocolate;
using HotChocolate.Data;
using HotChocolate.Execution;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;

namespace GraphQL.Tests;

public class BookTests
{
    [Fact]
    public async Task Book_Schema_Changed()
    {
        ISchema schema = await new ServiceCollection()
            .AddPooledDbContextFactory<LibraryDbContext>(
                options => options.UseInMemoryDatabase("Data Source=library.db"))
            .AddGraphQL()
            .RegisterDbContext<LibraryDbContext>(DbContextKind.Pooled)
            .AddMutationConventions()
            .AddQueryType()
            .AddMutationType()
            .AddSubscriptionType()
            .AddTypeExtension<BookQuery>()
            .AddTypeExtension<BookMutation>()
            .AddTypeExtension<BookSubscription>()
            .AddType<BookType>()
            .AddGlobalObjectIdentification()
            .AddFiltering()
            .AddSorting()
            .AddProjections()
            .BuildSchemaAsync();

        schema.Print().MatchSnapshot();
    }
}