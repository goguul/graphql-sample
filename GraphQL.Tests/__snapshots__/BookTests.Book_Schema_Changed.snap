schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

interface Error {
  message: String!
}

"The node interface is implemented by entities that have a global unique identifier."
interface Node {
  id: ID!
}

type AddBookPayload {
  book: Book
  errors: [AddBookError!]
}

type Author {
  id: ID!
  name: String!
  books: [Book!]!
}

type AuthorNotFoundError implements Error {
  message: String!
}

type Book implements Node {
  id: ID!
  title: String!
  authorId: ID!
  author: Author!
}

"A connection to a list of items."
type BooksConnection {
  "Information to aid in pagination."
  pageInfo: PageInfo!
  "A list of edges."
  edges: [BooksEdge!]
  "A flattened list of the nodes."
  nodes: [Book!]
}

"An edge in a connection."
type BooksEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Book!
}

type Mutation {
  addBook(input: AddBookInput!): AddBookPayload!
}

"Information about pagination in a connection."
type PageInfo {
  "Indicates whether more edges exist following the set defined by the clients arguments."
  hasNextPage: Boolean!
  "Indicates whether more edges exist prior the set defined by the clients arguments."
  hasPreviousPage: Boolean!
  "When paginating backwards, the cursor to continue."
  startCursor: String
  "When paginating forwards, the cursor to continue."
  endCursor: String
}

type Query {
  "Fetches an object given its ID."
  node("ID of the object." id: ID!): Node
  "Lookup nodes by a list of IDs."
  nodes("The list of node IDs." ids: [ID!]!): [Node]!
  books("Returns the first _n_ elements from the list." first: Int "Returns the elements in the list that come after the specified cursor." after: String "Returns the last _n_ elements from the list." last: Int "Returns the elements in the list that come before the specified cursor." before: String where: BookFilterInput order: [BookSortInput!]): BooksConnection
  bookById(id: ID!): Book!
}

type Subscription {
  bookAdded: Book!
}

union AddBookError = AuthorNotFoundError

input AddBookInput {
  book: BookInput!
}

input AuthorFilterInput {
  and: [AuthorFilterInput!]
  or: [AuthorFilterInput!]
  id: IdOperationFilterInput
  name: StringOperationFilterInput
  books: ListFilterInputTypeOfBookFilterInput
}

input AuthorInput {
  id: ID!
  name: String!
  books: [BookInput!]!
}

input AuthorSortInput {
  id: SortEnumType
  name: SortEnumType
}

input BookFilterInput {
  and: [BookFilterInput!]
  or: [BookFilterInput!]
  id: IdOperationFilterInput
  title: StringOperationFilterInput
  authorId: IdOperationFilterInput
  author: AuthorFilterInput
}

input BookInput {
  id: ID!
  title: String!
  authorId: ID!
  author: AuthorInput!
}

input BookSortInput {
  id: SortEnumType
  title: SortEnumType
  authorId: SortEnumType
  author: AuthorSortInput
}

input IdOperationFilterInput {
  eq: ID
  neq: ID
  in: [ID]
  nin: [ID]
}

input ListFilterInputTypeOfBookFilterInput {
  all: BookFilterInput
  none: BookFilterInput
  some: BookFilterInput
  any: Boolean
}

input StringOperationFilterInput {
  and: [StringOperationFilterInput!]
  or: [StringOperationFilterInput!]
  eq: String
  neq: String
  contains: String
  ncontains: String
  in: [String]
  nin: [String]
  startsWith: String
  nstartsWith: String
  endsWith: String
  nendsWith: String
}

enum SortEnumType {
  ASC
  DESC
}
